### ISSUE TITLE
PERF401 gives incorrect suggestions and requires suboptimal workarounds with async iterators

### ISSUE THREAD

**Issue by MarkusSintonen** [astral-sh/ruff#16315]
### Description

PERF401 triggers in a following case incorrectly and gives an incorrect suggestion. Making Ruff happy requires creating unnecessary temporary list. Which is not a great workaround.

```python
from collections.abc import AsyncIterator


async def my_nested_aiter() -> AsyncIterator[AsyncIterator[int]]:
    async def inner() -> AsyncIterator[int]:
        yield 1

    yield inner()


async def test_ruff_issue():
    my_list: list[int] = []
    async for inner_aiter in my_nested_aiter():
        async for x in inner_aiter:
            # Fails in Ruff with. "PERF401 Use `list.extend` with an async comprehension to create a transformed list"
            my_list.append(x)


async def test_incorrect_ruff_suggestion():
    my_list: list[int] = []
    async for inner_aiter in my_nested_aiter():
        # TypeError: 'async_generator' object is not iterable
        my_list.extend(x async for x in inner_aiter)


async def test_bad_work_around():
    my_list: list[int] = []
    async for inner_aiter in my_nested_aiter():
        # Works but requires unnecessary temporary lists for each iteration, just to get Ruff happy
        my_list.extend([x async for x in inner_aiter])

```

version:
```
% uv run ruff --version
ruff 0.9.3
```

**Comment by MarkusSintonen** [comment]
Related https://github.com/astral-sh/ruff/issues/11316#issuecomment-2114185025 but the original issue describes yet another issue with the rule as list extend does not work with async generator

**Comment by MarkusSintonen** [comment]
> The intended fix is probably this:
> 
> my_list = [
> 	x
> 	async for inner_aiter in my_nested_aiter()
> 	async for x in inner_aiter
> ]

I dont see how this would be any better. The nested comprehensions can get really spaghetti. The example I gave is also a very basic example. Also this has nothing to do with "PERF" rules performance-aspects.

> So I think ruff is actually giving a correct suggestion here

Yes it is correct but is it really performant as the "PERF" rules try to do, I would bet it is not. So the usage of temporary lists is really not a great workaround and is not inline with "PERF" rules. As it is not performance net positive.

> I also assume, like you and others in the linked issue, that creating a temporary list just to use extend is probably a net negative, so I'm a bit skeptical that this should apply to async iterators at all, at least in the extend case.

I would say the PERF401 should not be applied to async iterators at all whenever it tries to suggest usage of `list.extend`. Python does not support `extend` with async iterators so I would say it should not be applicable in the first place. None of the suggestions are performance related (ugly nested comprehensions), or performance net positive (temp lists).

**Comment by Tishka17** [comment]
It is not also not optimal for normal generators:
```python
def foo():
    a = []
    for x in range(100):
        if x % 2 == 0:
            a.append(x)
def bar():
    a = []
    a.extend(
        x for x in range(100) if x % 2 == 0
    )

n= 1000000
print(timeit.timeit("foo()", globals=globals(), number=n))
print(timeit.timeit("bar()", globals=globals(), number=n))
```

output:
1.9236854760092683
2.2944130940013565

I see the same on python 3.12 and python 3.14
generators are slower than normal for loop
